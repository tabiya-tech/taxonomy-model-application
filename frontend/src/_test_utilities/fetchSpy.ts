export function setupFetchSpy(
  expectedStatus: number,
  expectedResponseBody: string | object | undefined,
  contentType: "" | "application/json;charset=UTF-8"
): jest.SpyInstance {
  const responseBody =
    typeof expectedResponseBody === "string" ? expectedResponseBody : JSON.stringify(expectedResponseBody);
  const expectedResponse = new Response(responseBody, {
    status: expectedStatus,
    headers: { "Content-Type": contentType },
  });
  return jest.spyOn(window, "fetch").mockResolvedValue(expectedResponse);
}
