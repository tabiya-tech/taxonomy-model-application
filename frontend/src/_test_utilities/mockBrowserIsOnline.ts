import { fireEvent } from "@testing-library/react";

export function mockBrowserIsOnLine(isOnline: boolean) {
  if (window.navigator.onLine !== isOnline) {
    jest.spyOn(window.navigator, "onLine", "get").mockReturnValue(isOnline);
    if (isOnline) {
      fireEvent.online(window);
    } else {
      fireEvent.offline(window);
    }
  }
}

export function unmockBrowserIsOnLine() {
  jest.spyOn(window.navigator, "onLine", "get").mockRestore();
}
