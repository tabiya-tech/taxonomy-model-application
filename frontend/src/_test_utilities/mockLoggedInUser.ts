import { AuthContextValue, TabiyaUser, TabiyaUserRole, authContextDefaultValue } from "src/auth/AuthProvider";

export function mockLoggedInUser(mockedValues: Partial<AuthContextValue>){
  let user: TabiyaUser = mockedValues.user || { username: "foo", roles: [] };
  jest.spyOn(require("src/auth/hooks/useAuthUser"), "useAuthUser").mockReturnValue({
    ...authContextDefaultValue,
    user: user,
    setUser: (user: TabiyaUser) => null,
    hasRole: (role: TabiyaUserRole) => user.roles.includes(role),
    updateUserByAccessToken: () => {},
    ...mockedValues
  });
}
