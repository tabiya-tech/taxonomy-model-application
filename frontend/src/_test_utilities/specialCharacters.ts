const sampleValidSpecialCharacters = "^°!\"§$%&/()=?`´*+#'-_.,;<>öäü@[]{}|¡“¶¢[]≠¿'„…∞~<•±æœ";
const sampleInternationalCharacters = "αβγδεζηθικλμνξοπρστυφχψω";
const THREE_BYTE_UTF8_CHAR = "€";
export const WHITESPACE = " \n\r\t";

export function getTestString(length: number, prefix: string = ""): string {
  return (prefix + sampleValidSpecialCharacters + sampleInternationalCharacters)
    .slice(0, length)
    .padEnd(length, THREE_BYTE_UTF8_CHAR);
}

export function getRandomString(length: number) {
  let result = "";
  const characters = sampleValidSpecialCharacters + sampleInternationalCharacters;
  const charactersLength = characters.length;
  let counter = 0;
  while (counter < length) {
    result += characters.charAt(Math.floor(Math.random() * charactersLength)); // NOSONAR
    counter += 1;
  }
  return result;
}

export function getRandomLorem(length: number) {
  let result = "";
  const characters = "     abcdefghijklmnopqrstuvwxyz"; // 5 spaces for 20% chance of whitespace in the english language
  const charactersLength = characters.length;
  let counter = 0;
  while (counter < length) {
    const newChar = characters.charAt(Math.floor(Math.random() * charactersLength)); // NOSONAR
    if (!(newChar === " " && result.charAt(result.length))) {
      result += newChar;
      counter += 1;
    }
  }
  return result;
}
