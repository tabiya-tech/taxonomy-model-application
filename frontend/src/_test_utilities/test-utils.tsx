// Based on https://testing-library.com/docs/react-testing-library/setup/

import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@mui/material";
import applicationTheme, { ThemeMode } from "src/theme/applicationTheme/applicationTheme";
import SnackbarProvider from "src/theme/SnackbarProvider/SnackbarProvider";
import { IsOnlineProvider } from "src/app/providers";

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <IsOnlineProvider>
      <ThemeProvider theme={applicationTheme(ThemeMode.LIGHT)}>
        <SnackbarProvider>{children}</SnackbarProvider>
      </ThemeProvider>
    </IsOnlineProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
