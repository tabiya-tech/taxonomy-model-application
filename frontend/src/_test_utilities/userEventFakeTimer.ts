import userEvent from "@testing-library/user-event";
import { act } from "@testing-library/react";
import { UserEvent } from "@testing-library/user-event/setup/setup";

async function setup(callback: (user: UserEvent) => Promise<void>) {
  // We need to use fake times to ensure the debounce timer is progressed,
  // Since the debounce timer is called after the user stops typing, we need to set the fake timers
  // before the user begins typing, and for this reason we need to set up the userEvent to advance the timers using the jest timers,
  // so that the await user.type() can be resolved.
  // User enters a model name
  jest.useFakeTimers();
  const userEventFakeTimer = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
  await callback(userEventFakeTimer);
  act(() => jest.runOnlyPendingTimers());
  jest.useRealTimers();
}

export async function typeDebouncedInput(element: Element, text: string) {
  await setup(async (user) => {
    await user.type(element, text);
  });
}

export async function clickDebouncedButton(element: Element) {
  await setup(async (user) => {
    await user.click(element);
  });
}
