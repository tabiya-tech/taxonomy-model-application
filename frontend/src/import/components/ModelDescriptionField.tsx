import react from "react";
import {FormControl, FormLabel, Stack, TextareaAutosize} from "@mui/material";
import {generateUniqueId} from "src/utils/generateUniqueId";
import {useStyles} from "src/global.style";

export const TEXT = {
  MODEL_DESC_LABEL: "Model Description",
  MODEL_DESC_PLACEHOLDER: "Enter model description"
}
const baseTestID = "110643ab-593e-47dc-b398-ecc98518660e"

export const DATA_TEST_ID = {
  MODEL_DESC_INPUT: `model-desc-input-${baseTestID}`,
  MODEL_DESC_LABEL: `model-desc-label-${baseTestID}`
}

export interface ModelDescriptionFieldProps {
  notifyModelDescriptionChanged?: (newDescription: string) => void
}

export const ModelDescriptionField = ({notifyModelDescriptionChanged}: ModelDescriptionFieldProps) => {
  const uniqueId = generateUniqueId();

  function handleTextInputChange(e: react.ChangeEvent<HTMLTextAreaElement>) {
    if (notifyModelDescriptionChanged) {
      notifyModelDescriptionChanged(e.target.value);
    }
  }

  const classes = useStyles();
  return <FormControl sx={{width: '100%'}}>
    <Stack className={classes.fieldStack} spacing={0.5}>
      <FormLabel data-testid={DATA_TEST_ID.MODEL_DESC_LABEL}
                 htmlFor={uniqueId}>{TEXT.MODEL_DESC_LABEL}</FormLabel>
      <TextareaAutosize
        data-testid={DATA_TEST_ID.MODEL_DESC_INPUT}
        placeholder={TEXT.MODEL_DESC_PLACEHOLDER}
        id={uniqueId}
        minRows={6}
        style={{maxWidth: '100%', minWidth: '100%', maxHeight: '120px', minHeight: '100px'}}
        onChange={handleTextInputChange}
      />
    </Stack>
  </FormControl>
};
export default ModelDescriptionField;