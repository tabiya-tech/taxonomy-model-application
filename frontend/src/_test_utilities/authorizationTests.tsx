import { render, screen } from "src/_test_utilities/test-utils";
import AuthAPISpecs from "api-specifications/auth";
import { mockLoggedInUser } from "./mockLoggedInUser";

export const ALL_USERS = [
  AuthAPISpecs.Enums.TabiyaRoles.ANONYMOUS,
  AuthAPISpecs.Enums.TabiyaRoles.REGISTERED_USER,
  AuthAPISpecs.Enums.TabiyaRoles.MODEL_MANAGER,
];

type AuthorizationTestsParams = {
  name: string;
  Component: JSX.Element;
  testIds?: string[];
  roles?: AuthAPISpecs.Enums.TabiyaRoles[];
  tests?: Partial<
    Record<
      AuthAPISpecs.Enums.TabiyaRoles,
      {
        visible?: string[];
        hidden?: string[];
        visibleDisabled?: string[];
      }
    >
  >;
};

export const authorizationTests = {
  defaultName: "Test the rendering of the component based on Roles",
  callback: ({ name, Component, testIds = [], roles = ALL_USERS, tests }: AuthorizationTestsParams) => {
    return () => {
      test.each(roles?.map((role) => [role]))(`${name} renders correctly for user with role %s`, (role) => {
        const commonElements = testIds || [];
        const { visible = [], visibleDisabled = [], hidden = [] } = tests?.[role] || {};

        mockLoggedInUser({ user: { username: "foo", roles: [role] } });

        render(Component);

        commonElements.forEach((testId) => expect(screen.getByTestId(testId)).toBeInTheDocument());

        visible?.forEach((testId) => {
          expect(screen.getByTestId(testId)).toBeVisible();
        });

        hidden?.forEach((testId) => {
          expect(screen.queryByTestId(testId)).not.toBeInTheDocument();
        });

        visibleDisabled?.forEach((testId) => {
          const element = screen.getByTestId(testId);
          expect(element).toBeVisible();
          if (element instanceof HTMLInputElement) {
            expect(element).toBeDisabled();
          }
        });
      });
    };
  },
};
