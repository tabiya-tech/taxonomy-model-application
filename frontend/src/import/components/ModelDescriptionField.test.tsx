// mute the console
import "src/_test_utilities/consoleMock";

import { render, screen } from "src/_test_utilities/test-utils";
import ModelDescriptionField, { DATA_TEST_ID, TEXT } from "./ModelDescriptionField";
import { getTestString } from "src/_test_utilities/specialCharacters";
import { isSpecified } from "src/utils/isUnspecified";
import React from "react";
import { typeDebouncedInput } from "src/_test_utilities/userEventFakeTimer";

describe("ModelDescriptionField render tests", () => {
  beforeEach(() => {
    (console.error as jest.Mock).mockClear();
    (console.warn as jest.Mock).mockClear();
  });

  test("should render default state", () => {
    // WHEN the model name field is rendered
    render(<ModelDescriptionField />);

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();
    // AND the descriptionField to be visible
    const descField = screen.getByTestId(DATA_TEST_ID.MODEL_DESCRIPTION_FIELD);
    expect(descField).toBeInTheDocument();

    // AND expect the import to be visible
    const input = screen.getByTestId(DATA_TEST_ID.MODEL_DESC_INPUT);
    // AND expect the input to have a correct id
    expect(isSpecified(input.id)).toBe(true);
    // AND expect the input to have the empty string as value
    expect(input).toHaveValue("");
    // AND expect the placeholder to be correct
    expect(input).toHaveAttribute("placeholder", TEXT.MODEL_DESC_PLACEHOLDER);
    // AND expect label to be visible
    const label: HTMLLabelElement = screen.getByTestId(DATA_TEST_ID.MODEL_DESC_LABEL);
    expect(label).toBeInTheDocument();
    expect(label).toBeVisible();
    // AND expect label to have correct text
    expect(label).toHaveTextContent(TEXT.MODEL_DESC_LABEL);
    // AND expect label to be linked to inputField
    expect(label).toHaveAttribute("for", input.id);
  });

  test("multiple components should have a unique id", () => {
    // WHEN two model name fields are rendered
    render(<ModelDescriptionField />);
    render(<ModelDescriptionField />);

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();
    // AND the two inputs to have different ids
    const inputs = screen.getAllByTestId(DATA_TEST_ID.MODEL_DESC_INPUT);
    expect(inputs.length).toBe(2);
    expect(inputs[0].id).not.toBe(inputs[1].id);
  });
});

describe("ModelDescriptionField action tests", () => {
  test("should correctly notify the notifyModelDescriptionChanged handler", async () => {
    // GIVEN a notifyModelDescriptionChanged mock
    const notifyModelDescriptionChangedHandlerMock = jest.fn();

    // AND that the model description field is rendered
    render(<ModelDescriptionField notifyModelDescriptionChanged={notifyModelDescriptionChangedHandlerMock} />);

    // WHEN descriptionField changes value
    const givenModelDescription = getTestString(10);
    const descField = screen.getByTestId(DATA_TEST_ID.MODEL_DESC_INPUT);

    // We must use fake timers so that time can progress immediately while the input is debounced.
    await typeDebouncedInput(descField, givenModelDescription);

    // THEN expect the notifyModelDescriptionChangedHandlerMock to have been called once because it is debounced
    expect(notifyModelDescriptionChangedHandlerMock).toHaveBeenCalledTimes(1);

    // AND expect the notifyModelDescriptionChangedHandlerMock to have been called with the correct value
    expect(notifyModelDescriptionChangedHandlerMock).toHaveBeenCalledWith(givenModelDescription);
  });

  test("should handle text changes even if notifyModelNameChanged handler is notset", async () => {
    // GIVEN the model description field is rendered without a notifyModelDescriptionChanged handler
    render(<ModelDescriptionField />);

    // WHEN the descriptionField changes value
    const givenModelDescription = getTestString(10);
    const descField: HTMLInputElement = screen.getByTestId(DATA_TEST_ID.MODEL_DESC_INPUT);
    await typeDebouncedInput(descField, givenModelDescription);

    // THEN expect inputField to have the new value
    expect(descField.value).toBe(givenModelDescription);
  });
});
