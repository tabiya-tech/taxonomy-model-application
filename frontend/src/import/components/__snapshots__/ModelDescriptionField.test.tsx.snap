// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelDescriptionField render tests should render default state 1`] = `
<div
  class="MuiFormControl-root css-q8hpuo-MuiFormControl-root"
  data-testid="model-desc-field-110643ab-593e-47dc-b398-ecc98518660e"
>
  <div
    class="MuiStack-root css-ri2h3u-MuiStack-root"
  >
    <label
      class="MuiFormLabel-root MuiFormLabel-colorPrimary css-1k0yz6l-MuiFormLabel-root"
      data-testid="model-desc-label-110643ab-593e-47dc-b398-ecc98518660e"
      for="110643ab-593e-47dc-b398-ecc98518660e"
    >
      Model Description
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-multiline css-6chre-MuiInputBase-root-MuiInput-root"
      style="max-width: 100%; min-width: 100%; max-height: 20rem; min-height: 6rem;"
    >
      <textarea
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline css-18em216-MuiInputBase-input-MuiInput-input"
        data-testid="model-desc-input-110643ab-593e-47dc-b398-ecc98518660e"
        id="110643ab-593e-47dc-b398-ecc98518660e"
        placeholder="Enter model description"
        rows="10"
        style="height: 0px; overflow: hidden;"
      />
      <textarea
        aria-hidden="true"
        class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline css-18em216-MuiInputBase-input-MuiInput-input"
        readonly=""
        style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
        tabindex="-1"
      />
    </div>
  </div>
</div>
`;
